File: benchmarks/multilang-wasm/ml-gemm/gemm.dart
// ml-gemm Dart WasmGC kernel — mirrors benchmarks/v2/ml-gemm/workload.js:
// C = C0 + A * B, strict f32 left-to-right accumulation in frozen i/j/k order.
//
// Dart has no f32 primitive, so every multiply and every add is rounded with
// Math.fround (f64 -> f32) exactly like the JS oracle:
//   acc = fround(acc + fround(a * b))
// This is bit-identical to hardware f32 arithmetic (f32xf32 products are exact
// in f64; fround(acc + exactProduct) == f32.add). The fround overhead is real
// and disclosed per-variant. Outputs are written as fround(acc + 0) to
// normalize -0 to +0, matching the oracle's "acc + 0".

import 'dart:js_interop';

@JS('Math.fround')
external double fround(double x);

@JSExport()
class GemmKernels {
  @JSExport('gemm')
  void gemm(
    JSFloat32Array aJs,
    JSFloat32Array bJs,
    JSFloat32Array c0Js,
    JSFloat32Array outJs,
    int m,
    int n,
    int k,
  ) {
    final a = aJs.toDart; // zero-copy Float32List views on Wasm
    final b = bJs.toDart;
    final c0 = c0Js.toDart;
    final out = outJs.toDart;
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        double acc = c0[i * n + j];
        for (int t = 0; t < k; t++) {
          acc = fround(acc + fround(a[i * k + t] * b[t * n + j]));
        }
        out[i * n + j] = fround(acc + 0.0);
      }
    }
  }
}

void main() {
  dartKernels = createJSInteropWrapper(GemmKernels());
}

@JS('dartKernels')
external set dartKernels(JSObject value);

File: benchmarks/multilang-wasm/ml-gemm/gemm.ts
// gemm.ts — AssemblyScript multilang kernel for ml.gemm.v1.
//
// Same ABI and same arithmetic as gemm.c / gemm.rs: C = C0 + A * B with strict
// f32 left-to-right accumulation in the frozen ascending i/j/k order. Pointers
// are raw linear-memory byte offsets; no allocation, no runtime imports, so the
// engine can be instantiated with no import object.
//
// AssemblyScript's f32 arithmetic lowers to wasm f32.mul / f32.add, which
// rounds to f32 after every operation — bit-identical to the C and Rust
// kernels and to the JS oracle's Math.fround formulation.

export function gemm(
  a: usize,
  b: usize,
  c0: usize,
  out: usize,
  m: u32,
  n: u32,
  k: u32,
): void {
  for (let i: u32 = 0; i < m; i++) {
    for (let j: u32 = 0; j < n; j++) {
      let acc: f32 = load<f32>(c0 + <usize>(i * n + j) * 4);
      for (let t: u32 = 0; t < k; t++) {
        acc += load<f32>(a + <usize>(i * k + t) * 4) * load<f32>(b + <usize>(t * n + j) * 4);
      }
      // "acc + 0" normalizes -0 to +0, matching the JS oracle's "acc + 0".
      store<f32>(out + <usize>(i * n + j) * 4, acc + 0.0);
    }
  }
}
